### PR TITLE
Task/v2 paths

### LINKED ISSUE #259: Implement a first bunch of APIv2 operations
All the ones with the "Action" column filled at https://docs.google.com/spreadsheets/d/1f4m624nmO3jRjNalGE11lFLQixnfCENMV6dc54wUCCg/edit#gid=50130961, from `GET /v2/` to `DELETE	/v2/subscriptions/<subId>`.

### LINKED ISSUE #262: Distinguish NGSIv2 actions based on query parameters as well as URL
Some NGSIv2 actions must be distinguished by reading some query paremeters. E.g.:
```
POST/v2/entities/<entityId>
```
and 
```
POST /v2/entities/<entityId>?op=append
```
The Orion plugin has to be modified to allow the same distinctions.

### PR DESCRIPTION
**Pull request by dmoranj**
Fixes #259
Fixes #262

### CODE REVIEW

File: lib/plugins/orionUrls.js [hunk 1]
```diff
@@ -29,68 +29,86 @@
  * assigned.
  */
 module.exports = [
-  /* Standard NGSI operations */
```
Comment (fgalan) [inline lib/plugins/orionUrls.js]: Only indent changes in this fragment?

File: lib/plugins/orionUrls.js [hunk 2]
```diff
@@ -29,68 +29,86 @@
...
+    ['POST', /\/(v1\/registry|ngsi9)\/unsubscribecontextavailability$/, 'subscribe-availability'],
+    ['POST', /\/(v1|ngsi10)\/subscribecontext$/, 'subscribe'],
+    ['POST', /\/(v1|ngsi10)\/updatecontextsubscription$/, 'subscribe'],
+    ['POST', /\/(v1|ngsi10)\/unsubscribecontext$/, 'subscribe'],
+    ['POST', /\/(v1\/registry|ngsi9)\/registercontext$/, 'register'],
+    ['POST', /\/(v1\/registry|ngsi9)\/discovercontextavailability$/, 'discover'],
+    ['POST', /\/v1\/contexttypes$/, 'read'],
 
 
-  /* "Classic" NGSI9 operations */
```
Comment (fgalan) [inline lib/plugins/orionUrls.js]: Only indent changes in this fragment?

File: lib/plugins/orionUrls.js (hunk 1 above)
Comment (dmoranj) [inline lib/plugins/orionUrls.js]: Yes, I corrected the indentation in this file, but all the changes are bellow the `V2 Operations` tag.

File: lib/plugins/orionUrls.js (hunk 2 above)
Comment (dmoranj) [inline lib/plugins/orionUrls.js]: Same as above

File: lib/plugins/orionUrls.js (hunk 1 above)
Comment (fgalan) [inline lib/plugins/orionUrls.js]: NTC

File: lib/plugins/orionUrls.js (hunk 2 above)
Comment (fgalan) [inline lib/plugins/orionUrls.js]: NTC

File: lib/plugins/orionUrls.js [hunk 3]
```diff
@@ -29,68 +29,86 @@
...
+    ['GET', /^\/v1\/contextsubscriptions/, 'read'],
+    ['GET', /^\/v1\/contextsubscriptions\/.+/, 'read'],
+    ['GET', /^\/v1\/contexttypes/, 'read'],
+    ['GET', /^\/v1\/contexttypes\/.+/, 'read'],
+
+    /* V2 Operations */
+    ['GET', /^\/v2$/, 'read'],
+    ['GET', /^\/v2\/entities$/, 'read'],
+    ['GET', /^\/v2\/entities\/.+/, 'read'],
+    ['POST', /^\/v2\/entities$/, 'create', {op: 'append'}],
```
Comment (fgalan) [inline lib/plugins/orionUrls.js]: Not sure if at the end NGSIv2 uses `op=append` or `options=...,append,...`. I'm checking it...

File: README.md [hunk 4]
```diff
@@ -626,6 +626,25 @@ An up-to-date list of the convenience operations can be found [here](https://doc
...
 
+#### NGSIv2 
+| Method | Path                                                                                    | Action |
+| ------ |:--------------------------------------------------------------------------------------- | ---:|
+| GET    | /v2                                                                                     | R |
+| GET    | /v2/entities                                                                            | R |
+| GET    | /v2\/entities/{entityId}                                                                | R |
+| POST   | /v2\/entities                                                                           | C |
+| PATCH  | /v2\/entities/{entityId}                                                                | U |
+| POST   | /v2\/entities/{entityId}                                                                | U |
```
Comment (fgalan) [inline README.md]: Maybe an entry for `op=append` or `options=...,append,...` (whatever the correct one would be) is missing in the table?

Comment (fgalan) [inline README.md]: At the end, the right one is `options=...,append,...`.

Comment (dmoranj) [inline README.md]: Fixed in ba69d42

**Comment by anabelengp** [comment]
What about URLs related to subscriptions? In particular
`GET /v2/subscriptions` and `GET /v2/subscriptions/<subId>` that will be completely available in CB v0.25.0.
I though issue #259 was including them.

**Comment by dmoranj** [comment]
Fixed in 3223f92

**Comment by anabelengp** [comment]
LGTM

File: README.md (hunk 4 above)
Comment (fgalan) [inline README.md]: What about replacing

```
| POST   | /v2\/entities/{entityId}                                                                | U |
```

with

```
| POST   | /v2\/entities/{entityId}                                                                | U |
| POST   | /v2\/entities/{entityId}?options=append                                                 | C |
```

?

File: lib/plugins/orionUrls.js [hunk 5]
```diff
@@ -29,68 +29,88 @@
...
+    ['PUT', /^\/v2\/entities\/.+/, 'update'],
+    ['GET', /^\/v2\/entities\/.+\/attrs\/.+/, 'read'],
+    ['PUT', /^\/v2\/entities\/.+\/attrs\/.+/, 'update'],
+    ['DELETE', /^\/v2\/entities\/.+\/attrs\/.+/, 'delete'],
+    ['GET', /^\/v2\/entities\/.+\/attrs\/.+\/value$/, 'read'],
+    ['PUT', /^\/v2\/entities\/.+\/attrs\/.+\/value$/, 'update'],
+    ['GET', /^\/v2\/types$/, 'read'],
+    ['GET', /^\/v2\/types\/.+/, 'read'],
+    ['GET', /^\/v2\/subscriptions$/, 'read'],
+    ['GET', /^\/v2\/subscriptions\/.+/, 'read']
```
Comment (fgalan) [inline lib/plugins/orionUrls.js]: Being strict, #259 specifies "to DELETE /v2/subscriptions/<subId>", trying to "save time" in the future adding the rest of CRUD operations for subscriptions although currently they are not implemented. However, we can re-define the scope of #259 if you prefer to keep the code as it is now.

File: README.md [hunk 6]
```diff
@@ -626,6 +626,25 @@ An up-to-date list of the convenience operations can be found [here](https://doc
...
+| DELETE | /v2\/entities/{entityId}                                                                | D |
+| PUT    | /v2\/entities/{entityId}                                                                | U |
+| GET    | /v2\/entities/{entityId}/attrs/{attrId}                                                 | R |
+| PUT    | /v2\/entities/{entityId}/attrs/{attrId}                                                 | U |
+| DELETE | /v2\/entities/{entityId}/attrs/{attrId}                                                 | D |
+| GET    | /v2\/entities/{entityId}/attrs/{attrId}/value                                           | R |
+| PUT    | /v2\/entities/{entityId}/attrs/{attrId}/value                                           | U |
+| GET    | /v2\/types                                                                              | R |
+| GET    | /v2\/types/{typeId}                                                                     | R |
+
```
Comment (fgalan) [inline README.md]: What about subscriptions operations? Should they be included?

File: README.md (hunk 4 above)
Comment (dmoranj) [inline README.md]: Options are dealt with in the Orion plugin. The solution you mention was
the first approach, replaced to allow multiple options and quey parameters.

File: README.md (hunk 6 above)
Comment (dmoranj) [inline README.md]: You are right, I forgot to include them in the documentation. I'll include them.

**Comment by fgalan** [comment]
LGTM

**Pull request merged**